include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <map>
+#include <mutex>
+#include <rmm/detail/error.hpp>
+#include <rmm/detail/stack_trace.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <shared_mutex>
+#include <sstream>
+
+namespace rmm {
+namespace mr {
+/**
+ * @brief Resource that uses `Upstream` to allocate memory and tracks allocations.
+ *
+ * An instance of this resource can be constructed with an existing, upstream
+ * resource in order to satisfy allocation requests, but any existing allocations
+ * will be untracked. Tracking stores a size and pointer for every allocation, and a stack
+ * frame if `capture_stacks` is true, so it can add significant overhead.
+ * `tracking_resource_adaptor` is intended as a debug adaptor and shouldn't be used in
+ * performance-sensitive code. Note that callstacks may not contain all symbols unless
+ * the project is linked with `-rdynamic`. This can be accomplished with
+ * `add_link_options(-rdynamic)` in cmake.
+ *
+ * @tparam Upstream Type of the upstream resource used for
+ * allocation/deallocation.
+ */
+template <typename Upstream>
+class tracking_resource_adaptor final : public device_memory_resource {
+ public:
+  // can be a std::shared_mutex once C++17 is adopted
+  using read_lock_t  = std::shared_lock<std::shared_timed_mutex>;
+  using write_lock_t = std::unique_lock<std::shared_timed_mutex>;
+
+  /**
+   * @brief Information stored about an allocation. Includes the size
+   * and a stack trace if the `tracking_resource_adaptor` was initialized
+   * to capture stacks.
+   *
+   */
+  struct allocation_info {
+    std::unique_ptr<rmm::detail::stack_trace> strace;
+    std::size_t allocation_size;
+
+    allocation_info() : strace{nullptr}, allocation_size{0} {};
```

Comment:
I think default construction should be deleted. 
```suggestion
    allocation_info() = delete;
```